simpleline/base.py

```diff
@@ -19,682 +19,70 @@
-        self._handlers = {}
-
-        # screen stack contains triplets
-        #  UIScreen to show
-        #  arguments for it's refresh and setup method
-        #  value indicating whether new mainloop is needed
-        #   - None = do nothing
-        #   - True = execute new loop
-        #   - False = already running loop, exit when window closes
-        self._screens = []
-
-    def register_event_handler(self, event, callback, data=None):
-        """This method registers a callback which will be called when message "event"
-        is encountered during process_events.
-
-        The callback has to accept two arguments:
-        - the received message in the form of (type, [arguments])
-        - the data registered with the handler
-
-        :param event: the id of the event we want to react on
-        :type event: number|string
-
-        :param callback: the callback function
-        :type callback: func(event_message, data)
-
-        :param data: optional data to pass to callback
-        :type data: anything
-        """
-        if not event in self._handlers:
-            self._handlers[event] = []
-        self._handlers[event].append((callback, data))
-
-    def _thread_input(self, queue_instance, prompt, hidden):
-        """This method is responsible for interruptible user input.
-
-        It is expected to be used in a thread started on demand by the App class
-        and returns the input via the communication Queue.
-
-        :param queue_instance: communication queue_instance to be used
-        :type queue_instance: queue.Queue instance
-
-        :param prompt: prompt to be displayed
-        :type prompt: Prompt instance
-
-        :param hidden: whether typed characters should be echoed or not
-        :type hidden: bool
-        """
-        if hidden:
-            data = getpass.getpass(prompt)
-        else:
-            widget = TextWidget(str(prompt))
-            widget.render(self.width)
-            lines = widget.get_lines()
-            sys.stdout.write("\n".join(lines) + " ")
-            sys.stdout.flush()
-            # XXX: only one raw_input can run at a time, don't schedule another
-            # one as it would cause weird behaviour and block other packages'
-            # raw_inputs
-            if not RAW_INPUT_LOCK.acquire(False):
-                # raw_input is already running
-                return
-            else:
-                # lock acquired, we can run raw_input
-                try:
-                    data = input()
-                except EOFError:
-                    data = ""
-                finally:
-                    RAW_INPUT_LOCK.release()
-
-        queue_instance.put((hubQ.HUB_CODE_INPUT, [data]))
-
-    def switch_screen(self, ui, args=None):
-        """Schedules a screen to replace the current one.
+    __app = None
 
-        :param ui: screen to show
-        :type ui: instance of UIScreen
+    class AppPimpl(object):
```

Comment:
I'm used to use `pimpl` because it is a standard name for the Qt classes.

https://en.wikipedia.org/wiki/Private_class_data_pattern#Also_known_as

However, I don't have a problem to change this to some other name, if you don't want this shortcut.